My bowling object

Masaya-Nishijima opened the pull request.

Masaya-Nishijima committed:
feat: rbenvでrubyバージョンを指定

Masaya-Nishijima committed:
feat: RSpec Gemを追加

Masaya-Nishijima committed:
feat: RSpecセットアップ

hello.rb, hello_spec.rbはRSpecのテスト用プログラム

Masaya-Nishijima committed:
feat: rakeをインストール&Rakefileの作成

Masaya-Nishijima committed:
feat: RSpecの出力をドキュメント形式に変更

Masaya-Nishijima committed:
chore: テスト実行時に、各スペックの実行時間ランキングを表示させない

Masaya-Nishijima committed:
feat: shotの機能作成

Masaya-Nishijima committed:
build: rake 'new_rb[{file-name}]'で、libとsepcが自動作成。

codeコマンドを利用しているため、VScodeでセットアップしておく必要がある。

Masaya-Nishijima committed:
build: スペックの実行時間ログをコミットしないようにする。

Masaya-Nishijima committed:
feat: フレームの処理を作成

Masaya-Nishijima committed:
fix: RSpecの文言 typo 修正

Masaya-Nishijima committed:
feat: Gameクラス作成 スコアボード処理作成

Masaya-Nishijima committed:
feat: スコア計算システム完成(仮)

Masaya-Nishijima committed:
feat: 完了条件満すクラスを作成

Masaya-Nishijima committed:
refact: rubocopの規則に合せて修正

Masaya-Nishijima committed:
refactor: Gameクラスの外部から参照しないメソッドをプライベートに

Masaya-Nishijima committed:
test: 開発初期に必要と想定した動作が、不必要だったため削除

Masaya-Nishijima committed:
refactor: 変数名変更

Masaya-Nishijima committed:
feat: シェルコマンドとして実行できるようにmainを作成

Masaya-Nishijima committed:
test: 単体のスペックファイルでテストする際に必要なファイルの読み込みを追加

frame_spec.rbで、shot.rbを読み込むように修正

Masaya-Nishijima commented on 07.bowling_object/lib/game.rb:
```diff
@@ -0,0 +1,68 @@
+  attr_reader :score_board
+
+  def initialize(game_result)
+    @score_board = slice_result(game_result)
```
インスタンス変数の定義はこのように`initialize`内で行うか、直接`slice_result`内で行うかどちらが良いでしょうか?

若干コードは伸びてしまいますが、インスタンス変数という全体で使用するものなため、`initialize`での定義のほうが「明確にここで初期化してますよ!」と伝わるかと思い、ここで定義しています。

yuuu replied:
`@frames` への値のセットは `initialize` で行い、そのための処理は `slice_result` で行う、という方針自体は良いと思いました 👍 `score_board` を `attr_reader` していますが、この属性はクラスの外から参照する必要がないと思いますのでインスタンス変数ではなくローカル変数としたほうが良さそうですね。

Masaya-Nishijima replied:
確認ありがとうございます。


`score_board`の方はローカル変数に変更致しました。
[関連するスレッド](https://github.com/Masaya-Nishijima/ruby-practices/pull/9#discussion_r895020030)

Masaya-Nishijima commented on 07.bowling_object/main.rb:
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+require './lib/game'
+require './lib/frame'
+require './lib/shot'
```
```sh
$ ./main 6,3,9,0,0,3,8,2,7,3,X,9,1,8,0,X,6,4,5
139
```
のように使用するため`main.rb`で`shot`,`frame`,`game`をまとめて読み込んでいますが、
`main.rb`で`game`を読み込んで、`/lib/game.rb`で`frame`を読み込んで、`/lib/frame.rb`で`shot`を読み込むというマトリョーシカ構造とどちらが良いでしょうか?

マトリョーシカ構造の記述は、読み込みが連続してがなく気持ち悪い感じがします。
しかし、事実としてgame→frame→shotと依存している。

yuuu replied:
今回の場合、Gameクラスを使うためにはFrameクラスを使う必要があり、Frameクラスを使うためにはShotクラスを使う必要がある、という設計になっていますね。クラスの利用者にその依存関係は意識させないほうが良いので、`game.rb` の中で `frame` を、 `frame.rb` の中で `shot` を `require` したほうが良いです。

つまるところ、「ボウリングのスコア計算をしたい」という人は `game` のみを `require` して計算すれば良くて、中でどういったクラスを活用しているかは知る必要がないので。

Masaya-Nishijima replied:
>クラスの利用者にその依存関係は意識させないほうが良いので、game.rb の中で frame を、 frame.rb の中で shot を require したほうが良いです。

>つまるところ、「ボウリングのスコア計算をしたい」という人は game のみを require して計算すれば良くて、中でどういったクラスを活用しているかは知る必要がないので。

なるほど、こういう考え方で、どういうふうに`require`するか決めるのですね。勉強になります。

Masaya-Nishijima reviewed (commented):
レビュー前に自分で気になっている箇所

Masaya-Nishijima commented:
## プログラム全体の 疑問・問題

### 1. 変数名の使い分けが、たぶん自分以外には伝わらない。
変数命名の問題で、
スコア関係で
- `score`
- `point`

の二種類を使っている件について、


- `score`は、ストライクはX(char)、gameでは文字どおりスコアとして利用
- `point`は、ストライクのポイントは10(int)、また、全体を通してストライクやスペアの加算が行われていない素のスコアを表わしている。

と使い分けています。

### 2. gameクラスに処理が集中している。
- 純粋にコード量・複雑さが、`Shot : Frame : Game = 1 : 2 : 7`くらいになっている。
- 9フレーム目ストライクの処理のため、game内で、`frames[9].first_shot.point`など、frameクラス(`.first_shot`) → shotクラス(`.point`)と深くまで呼び出してしまっている。

Frameクラスで、ストライク、スペア処理を記述できないかな?
と思いましたが、Frameクラスのインスタンス同士が密に繋ってしまうため断念しました。

yuuu commented on 07.bowling_object/lib/game.rb:
```diff
@@ -0,0 +1,68 @@
+    @score_board = slice_result(game_result)
+
+    @frames = []
+    @score_board.each do |frame_score|
```
ここは `map` を使ったほうが良いですね

yuuu replied:
[ここ](https://github.com/Masaya-Nishijima/ruby-practices/pull/9/files#diff-ce106ea3ed6d4a253d9b2bbbe2288144a2751f26cc6b1988faa1baeb7bed82ffR8) も同様ですね。

Masaya-Nishijima replied:
2箇所とも変更致しました。
eachメソッドの汎用性の高さで、mapメソッドを利用することを忘れていました。🙇

yuuu commented on 07.bowling_object/lib/game.rb:
```diff
@@ -0,0 +1,68 @@
+      result_shot == 'X' ? 'X' : result_shot.to_i
+    end
+    (MAX_FRAME - 1).times do
+      answer << if result_array[0] == 'X'
```
三項演算子を使ってコードの重複をなくしましょう。

```ruby
answer = result_array.shift(result_array[0] == 'X' ? 1 : 2)
```

Masaya-Nishijima replied:
変更致しました。

メソッドの引数でも三項演算子は利用できたのですね(驚愕)。
今後のコーディングで利用していきます!

yuuu commented on 07.bowling_object/lib/game.rb:
```diff
@@ -0,0 +1,68 @@
+
+  def slice_result(result)
+    answer = []
+    result_array = result.split(',').map do |result_shot|
```
Rubyの場合、配列を示す変数名は複数形とするカルチャーがあるので `results` の方が良いですね。

Masaya-Nishijima replied:
変更致しました。`result_array`→`results`

確認ですが、`result`ではなく`result_array `を`results`にするという解釈で良いですか。
##
#### `slice_result`メソッドでの各変数の役割
result: 引数(配列ではなく、文字列でスコアが入力される)
result_array(→results): 中間の値(数値or'X'の一次元配列)
anser: 戻り値(二次元配列で`[[1フレーム1投目, 2投目], [2フレーム1投目, 2投目]...]`とスコアが記録される)

yuuu replied:
はい、合っています 👍

yuuu commented on 07.bowling_object/lib/game.rb:
```diff
@@ -0,0 +1,68 @@
+
+  def calc_strike_score
+    strike_score = 0
+    @frames.each_with_index do |frame, index|
```
ここは `@frames.each_with_index.sum` とすることで `strike_score` のお世話をしなくて良くなります。

```ruby
@frames.each_with_index.sum do |frame, index|
  # 各フレームのスコアを戻り値とする
end
```

Masaya-Nishijima replied:
sumにnilが入るとエラーが出るため、
- 10フレーム目の`next if index == MAX_FRAME - 1`を、`if (index == MAX_FRAME - 1) 0 else`と置き換えたためifのインデントが1段深まる。
- ストライク以外のときのため、`else 0 end`追加する。

と変更しましたが、構造が複雑すぎると、rubocopが通らない状態です。
```ruby
# 元のコード
 def calc_strike_score
    strike_score = 0
    @frames.each_with_index do |frame, index|
      next if index == MAX_FRAME - 1

      next_frame = @frames[index + 1]
      after_next_frame = @frames[index + 2]
      if frame.status == 'STRIKE'
        if next_frame.status == 'STRIKE'
          strike_score += next_frame.point
          strike_score += after_next_frame.first_shot.point if index != MAX_FRAME - 2
        else
          strike_score += index == 8 ? next_frame.first_shot.point + next_frame.second_shot.point : next_frame.point
        end
      end
    end
    strike_score
  end
```
```ruby
# strike_score変数を無くすため、内部が複雑になってしまったコード
 def calc_strike_score
    @frames.each_with_index.sum do |frame, index|
      if index == MAX_FRAME - 1
        0
      else
        next_frame = @frames[index + 1]
        after_next_frame = @frames[index + 2]
        if frame.status == 'STRIKE'
          if next_frame.status == 'STRIKE'
            next_frame.point + (index != MAX_FRAME - 2 ? after_next_frame.first_shot.point : 0)
          else
            index == 8 ? next_frame.first_shot.point + next_frame.second_shot.point : next_frame.point
          end
        else
          0
        end
      end
    end
  end
```

`@frames.each_with_index.map(&:to_i).sum`
, or `@frames.each_with_index.to_i.sum`
のようにnilを0にする処理(to_i)を外側で追加し、内部で0を出力するためだけのifのインデントや、elseを減らしたいのですが以下のようなエラーが出て動作しませんでした。
この場合どように記述すればto_iメソッドが利用できるのでしょうか?
```ruby
1) 完了条件 0,10,1,5,0,0,0,0,X,X,X,5,1,8,1,0,4
     Failure/Error:
       @frames.each_with_index.map(&:to_i).sum do |frame, index|
         if index == MAX_FRAME - 1
           nil
         else
           next_frame = @frames[index + 1]
           after_next_frame = @frames[index + 2]
           if frame.status == 'STRIKE'
             if next_frame.status == 'STRIKE'
               next_frame.point + (index != MAX_FRAME - 2 ? after_next_frame.first_shot.point : 0)
             else
     
     NoMethodError:
       undefined method `to_i' for #<Frame:0x000000011332d278 @shots=[#<Shot:0x000000011332d1b0 @point=0>, #<Shot:0x000000011332d160 @point=10>], @first_shot=#<Shot:0x000000011332d1b0 @point=0>, @second_shot=#<Shot:0x000000011332d160 @point=10>>
     
           @frames.each_with_index.map(&:to_i).sum do |frame, index|
                                  ^^^^
       Did you mean?  to_s
     # ./lib/game.rb:31:in `each'
     # ./lib/game.rb:31:in `each_with_index'
     # ./lib/game.rb:31:in `each'
     # ./lib/game.rb:31:in `map'
     # ./lib/game.rb:31:in `calc_strike_score'
     # ./lib/game.rb:15:in `score'
     # ./spec/sample_spec.rb:17:in `block (2 levels) in <top (required)>'
```
```ruby
  1) ゲーム スコアの算出 ストライクあり(連続ストライク、三投はなし) X,4,5,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
     Failure/Error:
       @frames.each_with_index.to_i.sum do |frame, index|
         if index == MAX_FRAME - 1
           nil
         else
           next_frame = @frames[index + 1]
           after_next_frame = @frames[index + 2]
           if frame.status == 'STRIKE'
             if next_frame.status == 'STRIKE'
               next_frame.point + (index != MAX_FRAME - 2 ? after_next_frame.first_shot.point : 0)
             else
     
     NoMethodError:
       undefined method `to_i' for #<Enumerator: [#<Frame:0x000000010cccb548 @shots=[#<Shot:0x000000010cccb480 @point=10>], @first_shot=#<Shot:0x000000010cccb480 @point=10>, @second_shot=nil>, #<Frame:0x000000010cccb430 @shots=[#<Shot:0x000000010cccb3e0 @point=4>, #<Shot:0x000000010cccb3b8 @point=5>], @first_shot=#<Shot:0x000000010cccb3e0 @point=4>, @second_shot=#<Shot:0x000000010cccb3b8 @point=5>>, #<Frame:0x000000010cccb390 @shots=[#<Shot:0x000000010cccb340 @point=0>, #<Shot:0x000000010cccb318 @point=0>], @first_shot=#<Shot:0x000000010cccb340 @point=0>, @second_shot=#<Shot:0x000000010cccb318 @point=0>>, #<Frame:0x000000010cccb2f0 @shots=[#<Shot:0x000000010cccb2a0 @point=0>, #<Shot:0x000000010cccb278 @point=0>], @first_shot=#<Shot:0x000000010cccb2a0 @point=0>, @second_shot=#<Shot:0x000000010cccb278 @point=0>>, #<Frame:0x000000010cccb250 @shots=[#<Shot:0x000000010cccb200 @point=0>, #<Shot:0x000000010cccb1d8 @point=0>], @first_shot=#<Shot:0x000000010cccb200 @point=0>, @second_shot=#<Shot:0x000000010cccb1d8 @point=0>>, #<Frame:0x000000010cccb1b0 @shots=[#<Shot:0x000000010cccb160 @point=0>, #<Shot:0x000000010cccb138 @point=0>], @first_shot=#<Shot:0x000000010cccb160 @point=0>, @second_shot=#<Shot:0x000000010cccb138 @point=0>>, #<Frame:0x000000010cccb110 @shots=[#<Shot:0x000000010cccb0c0 @point=0>, #<Shot:0x000000010cccb098 @point=0>], @first_shot=#<Shot:0x000000010cccb0c0 @point=0>, @second_shot=#<Shot:0x000000010cccb098 @point=0>>, #<Frame:0x000000010cccb070 @shots=[#<Shot:0x000000010cccb020 @point=0>, #<Shot:0x000000010cccaff8 @point=0>], @first_shot=#<Shot:0x000000010cccb020 @point=0>, @second_shot=#<Shot:0x000000010cccaff8 @point=0>>, #<Frame:0x000000010cccafd0 @shots=[#<Shot:0x000000010cccaf80 @point=0>, #<Shot:0x000000010cccaf58 @point=0>], @first_shot=#<Shot:0x000000010cccaf80 @point=0>, @second_shot=#<Shot:0x000000010cccaf58 @point=0>>, #<Frame:0x000000010cccaf30 @shots=[#<Shot:0x000000010cccaee0 @point=0>, #<Shot:0x000000010cccaeb8 @point=0>], @first_shot=#<Shot:0x000000010cccaee0 @point=0>, @second_shot=#<Shot:0x000000010cccaeb8 @point=0>>]:each_with_index>
     
           @frames.each_with_index.to_i.sum do |frame, index|
                                  ^^^^^
       Did you mean?  to_s
                      to_a
                      to_h
     # ./lib/game.rb:31:in `calc_strike_score'
     # ./lib/game.rb:15:in `score'
     # ./spec/game_spec.rb:23:in `block (4 levels) in <top (required)>'

```

Masaya-Nishijima replied:
rubocopエラー
```ruby
MacLocal: ~/Desktop/Fio/ruby-practices/07.bowling_object (my-bowling_object *=) 
$ rubocop lib   
Inspecting 3 files
.C.

Offenses:

lib/game.rb:31:3: C: Metrics/PerceivedComplexity: Perceived complexity for calc_strike_score is too high. [10/8]
  def calc_strike_score ...
  ^^^^^^^^^^^^^^^^^^^^^
lib/game.rb:40:33: C: Metrics/BlockNesting: Avoid more than 3 levels of block nesting.
            next_frame.point + (index != MAX_FRAME - 2 ? after_next_frame.first_shot.point : 0)
                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/game.rb:42:13: C: Metrics/BlockNesting: Avoid more than 3 levels of block nesting.
            index == 8 ? next_frame.first_shot.point + next_frame.second_shot.point : next_frame.point
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

3 files inspected, 3 offenses detected

yuuu replied:
@Masaya-Nishijima 
検討ありがとうございます。`@frames.each_with_index.map(&:to_i)` のような方法は `Frame#to_i` がないと動作しないですし、`Frame` を `Integer` に変更することはできない＆本末転倒だと思うので一旦忘れた方が良さそうです。

アドバイスするとしたら `next` でも戻り値を返せるので、これを使って基本的に元のコードのままとするのはいかがでしょうか？元のコードから `strike_score +=` が消えただけなのでむしろシンプルになると思っています。

```ruby
def calc_strike_score
  @frames.each_with_index.sum do |frame, index|
    next 0 if index == MAX_FRAME - 1

    next_frame = @frames[index + 1]
    after_next_frame = @frames[index + 2]
    if frame.status == 'STRIKE'
      if next_frame.status == 'STRIKE'
        next_frame.point + (index != MAX_FRAME - 2 ? after_next_frame.first_shot.point : 0)
      else
        index == 8 ? next_frame.first_shot.point + next_frame.second_shot.point : next_frame.point
      end
    else
      0
    end
  end
end
```

そのフレームがストライクでない場合にもearly returnならぬearly nextをすればよりスッキリしそうですね。

```ruby
def calc_strike_score
  @frames.each_with_index.sum do |frame, index|
    next 0 if index == MAX_FRAME - 1
    next 0 unless frame.status == 'STRIKE'

    next_frame = @frames[index + 1]
    after_next_frame = @frames[index + 2]
    
    if next_frame.status == 'STRIKE'
      next_frame.point + (index != MAX_FRAME - 2 ? after_next_frame.first_shot.point : 0)
    else
      index == 8 ? next_frame.first_shot.point + next_frame.second_shot.point : next_frame.point
    end
  end
end
```

yuuu replied:
`calc_spare_score` も同様に対処をお願いします 🙏

Masaya-Nishijima replied:
> `next`でも戻り値を返せる

こちら初めて知りました。(nextしたら絶対にnilが帰ってしまうものかと。Ruby快適すぎる。)
アドバイスありがとうございます。

Masaya-Nishijima replied:
すこしだけ、early nextの記述を変えていますが、ほぼ、提案していただいた形に修正しています。
```ruby
def calc_strike_score
  @frames.each_with_index.sum do |frame, index|
    next 0 if frame.status != 'STRIKE' || index == MAX_FRAME - 1

    next_frame = @frames[index + 1]
    after_next_frame = @frames[index + 2]
    
    if next_frame.status == 'STRIKE'
      next_frame.point + (index != MAX_FRAME - 2 ? after_next_frame.first_shot.point : 0)
    else
      index == 8 ? next_frame.first_shot.point + next_frame.second_shot.point : next_frame.point
    end
  end
end
```
変更点
1. 9フレーム目でearly nextすることよりも、ストライクではなくてearly nextすることの方が多いから、ストライクじゃないかの確認が先に。(感覚の問題と、処理効率の問題から)
2. 同じ`next 0`なので`||`でまとめて記述

yuuu commented on 07.bowling_object/lib/frame.rb:
```diff
@@ -0,0 +1,24 @@
+  end
+
+  def status
+    return 'THREE_SHOT' unless @shots[2].nil?
```
3番目の投球もインスタンス変数としたほうが、その値が何を示しているかわかりやすいですし、1投目と2投目とのレベル感も合うので良さそうです。

Masaya-Nishijima replied:
'THREE_SHOT'のステータスは、不要となっていたので削除しました。
```ruby
  def status
    if @shots[2].nil?
      return 'STRIKE' if @first_shot.point == 10
      return 'SPARE' if @shots.slice(0, 2).map(&:point).sum == 10
    end
  end
```

3投目のスコアは、計算に不必要なので追加していませんが、いかがでしょうか?


1投目のスコア:ストライクが連続したときに、2フレーム先の一投目を参照する必要がある。
2投目のスコア:9フレーム目でストライクでなおかつ、10フレームが3投になっているときにframe.pointをするとストライクの加算が2投ではなく3投になってしまうため。(一投目と合せて使う)

yuuu replied:
3投目のスコアは計算に不必要とのことですが `@shots[2].nil?` のコードは残っていますね。これを読んだときに 「 `@shots[2]` って何が格納されているんだっけ？」と疑問に思うことを考えると、 `@third_shot` という変数にしておいた方が可読性は高いと思います。

Masaya-Nishijima replied:
>3投目のスコアは計算に不必要とのことですが `@shots[2].nil?` のコードは残っていますね。

完全に認識外でした。(Gameクラスで、呼び出すかどうかだけ考えて、Frameクラス自身で呼び出すかを考慮していませんでした。)

Masaya-Nishijima replied:
775dc6e こちらで修正しています。

yuuu commented on 07.bowling_object/lib/frame.rb:
```diff
@@ -0,0 +1,24 @@
+    shots.each do |shot|
+      @shots.push Shot.new(shot)
+    end
+    @first_shot = @shots[0]
```
多重代入を使うとスッキリしますね 👍 

```ruby
@first_shot, @second_shot, @third_shot = shots.map { |shot| Shot.new(shot) }
```

Masaya-Nishijima replied:
[上コメント](https://github.com/Masaya-Nishijima/ruby-practices/pull/9#discussion_r895290246)のように`@third_shot`は不要なので`@first_shot, @second_shot`だけ代入するようにしています。

```ruby
  def initialize(shots)
    @first_shot, @second_shot = @shots = shots.map { |shot| Shot.new(shot) }
  end

  def point
    @shots.map(&:point).sum
  end
```

フレイムの主体は`@shots(Shotクラスのインスタンス配列)`で、`@first_shot, @second_shot (, @third_shot)`はあくまで、各ショットを直接参照できるラッパーと考えているので
`@first_shot, @second_shot, @third_shot = shots.map { |shot| Shot.new(shot) }`とし、`@shots`を無くしてはいませんがいかがでしょうか。

- `@shots`が無い場合、上の`pointメソッド`が
`if @third_shot.nil ? @first_shot.point + @second_shot.point : @first_shot.point + @second_shot.point +  @third_shot.point`のように冗長になってしまうデメリット。

- 仕様変更で、新ルールの計算をするプログラムにするなど、他フレームの一投目や二投目を参照する必要がなくなった場合に、本体の`@shots`だけを残し、ラッパーの`@first_shot, @second_shot (, @third_shot)`を削除することができる。

yuuu replied:
こちらは、 [上のコメント](https://github.com/Masaya-Nishijima/ruby-practices/pull/9#discussion_r897331467) を対処するのであればやはり `@shots` は不要になると思っています。

Masaya-Nishijima replied:
775dc6e こちらで修正しています。

Masaya-Nishijima replied:
<img width="1379" alt="スクリーンショット 2022-06-15 16 14 06" src="https://user-images.githubusercontent.com/100653079/173765822-26835d28-215e-4b5d-80fe-550a5694e6fb.png">

yuuu commented on 07.bowling_object/spec/game_spec.rb:
```diff
@@ -0,0 +1,79 @@
+    context 'ストライクがない場合' do
+      it '1,1,2,2,3,3,4,4,5,5,6,0,7,0,8,0,9,0,0,0' do
+        game = Game.new('1,1,2,2,3,3,4,4,5,5,6,0,7,0,8,0,9,0,0,0')
+        expect(game.score_board).to eq [[1, 1], [2, 2], [3, 3], [4, 4], [5, 5], [6, 0], [7, 0], [8, 0], [9, 0], [0, 0]]
```
レビュー後に提出物のコメントを読んだので、追加でコメントします 🙏 
仮に「Gameクラスのインスタンスは `score_board` というメソッドで、投球の二次元配列を取得できる」という仕様でクラスを作っているのであればこのテストも `score_board` の `attr_reader` も必要だと思います。

実際今回必要なのは、あくまで「スコアの計算」のみなので `文字列の入力をフレームごと配列化` のテストは不要ではないかと思っています。

今の方針だと、例えば「もっと効率の良い計算方法を見つけたので、文字列の入力をフレームごとに配列化せず、スコア計算するようにしたい」というケースが出たときに、クラスの入出力仕様は変わらないはずなのにテストを修正する必要が生じてしまいますね。

Masaya-Nishijima replied:
わざわざ追加でのコメントありがとうございます。
> レビュー後に提出物のコメントを読んだので、追加でコメントします 🙏
##
なるほど、了解致しました。(`文字列の入力をフレームごと配列化`のテストを削除しました。)
> 実際今回必要なのは、あくまで「スコアの計算」のみなので`文字列の入力をフレームごと配列化`のテストは不要ではないかと思っています。

今回の課題の作成にあたり、取った解法として、「1.フレームに二次元配列で分ける 2.計算処理をする」という手順で取り組んでいました。また、テスト駆動開発をしていたため、「1.フレームに二次元配列で分ける」のテストが完成していたため、消す勇気がなく消せていませんでした。

Masaya-Nishijima committed:
refactor: 各ソースで、依存するファイルを直接読み込むように変更

Masaya-Nishijima committed:
test: 要件にない仕様のテストを削除

Masaya-Nishijima committed:
refactor: Mapメソッドの活用

Masaya-Nishijima committed:
refactor: 三項演算子の活用

Masaya-Nishijima committed:
refactor: 配列名を複数形に

Masaya-Nishijima committed:
refactor?: cal_strike_scoreの変数をなくす

ifのインデントが深まる

Masaya-Nishijima committed:
refactor: 未使用のステータスTHREE_SHOTを削除

Masaya-Nishijima committed:
refactor: 多重代入の利用

Masaya-Nishijima committed:
refactor: rubocop(未修正有り)

Masaya-Nishijima committed:
refactor: ストライク、スペアの処理を変更

https://github.com/Masaya-Nishijima/ruby-practices/pull/9#discussion_r897324409

Masaya-Nishijima committed:
refactor: インスタンス変数:サードショットを作成
